apigee/registry
# Implement conformance command to generate Conformance Report based on API style guides

## Summary
The conformance command should compute conformance reports for specs that are provided according to all the API Style Guides that are specified on the project containing the spec. Conformance reports are defined by [this proto](https://github.com/apigee/registry/blob/c8b9dcca1354c28379a48b56effc11683aafad4e/google/cloud/apigee/registry/applications/v1alpha1/registry_conformance_report.proto). This pull request aims to implement this functionality. There is parallel computation of conformance reports when the spec provided is a wildcard that matches a range of specs. In this case, conformance reports for multiple specs are computed in parallel.

## Test
- Have registry running on `localhost:8080`.
- Ensure that https://github.com/APIs-guru/openapi-directory is extracted to your Desktop. You may delete multiple directories within the `APIs` directory, but keep `billbee.io`.
- Ensure that the following style guide is located under `~/styleguide.yaml`:
```yaml
id: openapistyleguide
mime_types:
  - application/x.openapi+gzip;version=2
  - application/x.openapi+gzip;version=3
guidelines:
  - id: refproperties
    display_name: Govern Ref Properties
    description: This guideline governs properties for ref fields on specs.
    rules:
      - id: norefsiblings
        description: An object exposing a $ref property cannot be further extended with additional properties.
        linter: spectral
        linter_rulename: no-$ref-siblings
        severity: 1
    status: 2
linters:
  - name: spectral
    uri: https://github.com/stoplightio/spectral
```

Now, run the following series of bash commands:
```bash
apg registry delete-project --name projects/example

apg registry create-project --project_id example --insecure --address localhost:8080

registry upload bulk openapi --project_id example ~/Desktop/openapi-directory/APIs

registry upload styleguide ~/styleguide.yaml --project_id=example

registry compute conformance projects/example/locations/global/apis/billbee.io/versions/v1/specs/swagger.yaml     

registry list projects/example/locations/global/apis/billbee.io/versions/v1/specs/swagger.yaml/artifacts/-
```

**Expected output**:
```bash
projects/example/locations/global/apis/billbee.io/versions/v1/specs/swagger.yaml/artifacts/conformance-openapistyleguide
```


The contents can be inspected via: `go run main.go get projects/example/locations/global/apis/billbee.io/versions/v1/specs/swagger.yaml/artifacts/conformance-openapistyleguide --contents`
